src/Components/Components/src/Routing/RoutingStateProvider.cs

```diff
@@ -0,0 +1,15 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+namespace Microsoft.AspNetCore.Components.Routing;
+
+/// <summary>
+/// Provides RouteData
+/// </summary>
+public class RoutingStateProvider
+{
+    /// <summary>
+    /// Gets RouteData
+    /// </summary>
+    public virtual RouteData? RouteData { get { return null; } }
+}
```

Comment:
This approach of `RouteData` being a property is very specialized to the `EndpointHtmlRendering` scenario, since it relies on the `RouteData` value not changing over time in reaction to navigation in advance of the router getting the value. (I know a custom getter could construct the value dynamically, but that wouldn't be idiomatic)

However I think we could quite easily make this into something more general-purpose. If it was changed to:

```cs
public interface IRouteDataProvider
{
    RouteData? GetRouteData();
}
```

... then the Endpoints pipeline could register a version that behaves equivalently to what we already have here, but also people doing interactive rendering could optionally make a version to does arbitrary route URL parsing and selecting page components. That would automatically cover some other feature requests:

 * https://github.com/dotnet/aspnetcore/issues/45396
 * https://github.com/dotnet/aspnetcore/issues/45653 (maybe, didn't check super carefully)
 * https://github.com/dotnet/aspnetcore/issues/42431
 * https://github.com/dotnet/aspnetcore/issues/40736

Given how many different requests we've had for customizing the URL matching, this seems like a big bonus to get for a small design update!

Reply:
I didn't think about this scenario. Thanks @SteveSandersonMS ! I'll change it.

Reply:
I started implementing the design with `RouteData? GetRouteData(string url);` and realized that it is not quite the same as the previous design as it may seem.
Previous `RoutingStateProvider` is a scoped service that provides `RouteData` with the component type and the route parameters. Example: there is `Counter.razor` page with route "/counter/{param1}/{param2}". if you go to "/counter/test1/test2" `RoutingStateProvide` will provide `RouteData` `PageType = typeof(Counter)` and `RouteValues = {{"param1", "test1"}, {"param2", "test2"}}`. The purpuse of `RoutingStateProvider` is to provide the `RouteData` to the `Router` component so the routing wouldn't happen twice as it already happened because of endpoints.
`IRouteDataProvider` in my understanding is a singleton service that is needed for mapping routes to component type in order to override the existing routes of the pages. (and also because of that the name should be different)
Hence we cannot use `IRouteStateProvider` instead of `RoutingStateProvider` as they play different roles. If we want add a feature to map routes to a page which route is different then it is a seperate thing and we still need `RoutingStateProvider`.

Reply:
> IRouteDataProvider in my understanding is a singleton service

I was expecting it to be a scoped service, and to do exactly what you've said that `RoutingStateProvider` would do, i.e.:

 * The default implementation would either:
   *  Just return `null` in all cases, if we're leaving `Router` to do the default `RouteData` creation
   *  ... or would contain the routedata-creation logic that currently lives in `Router`, if we're moving it into the service
 * The implementation used by `EndpointHtmlRenderer` would get routedata from what was already determined by ASP.NET Core's server-side routing system

Technically, `GetRouteData` would not need to take a `url` parameter, since implementations could depend on `NavigationManager` and get the URL from there. I only suggested having the `url` parameter since it would be easier for people who are trying to implement something like "if the URL is `/{lang}/{path}`, strip off `/{lang}` for localization and then call the default implementation passing `/{path}`".

Does this make sense? Sorry if I'm still missing something about why `IRouteDataProvider` couldn't have a role like this!

Reply:
We just discussed this, and you clarified what was missing from my thinking. You are right that:

 * It doesn't really make sense for a centralized service to do the `url`->`RouteData` mapping on its own without having access to the table of routes declared on all the `@page` directives everywhere. Because how would the central service know how each page wants to extract parameters from the route? That information only exists in the table of routes.
   * So, people who implement `IRouteDataProvider` would almost certainly want to fall back on the default routing behavior that consults the table of routes.
 * But it **does** make sense for people to want to do URL rewriting, i.e., to perform some translation on the URL string before it goes into the default `url`->`RouteData` mapping based on the table of routes.

Looking back at the list of issues I cited above:

 * https://github.com/dotnet/aspnetcore/issues/45396: it would be sufficient to support URL rewriting
 * https://github.com/dotnet/aspnetcore/issues/45653: it would be sufficient to support URL rewriting
 * https://github.com/dotnet/aspnetcore/issues/42431: no, this is unrelated - it's about mutating the global route table somehow
 * https://github.com/dotnet/aspnetcore/issues/40736: it would be sufficient to support URL rewriting (and in fact this is an identical request to the second item in this list)

So I think there are a few options here:

 * Forget about this extra feature - just have some `RouteDataProvider` that takes no parameters and is just a way for `Endpoints` to override what `Router` does.
   * We could separately implement some kind of URL rewriting feature in the future
 * Or, move the default `Router`'s route selection logic into a new `DefaultRouteDataProvider` that has a method like `public virtual RouteData GetRouteData(string url)` so people can subclass it as a way to do URL rewriting *and* we can have a version of it in Endpoints that ignores `url` and returns whatever `RouteData` was determined by the ASP.NET Core router.

Either approach would be fine in my opinion. Perhaps you could judge whether doing the second of these options would be time-consuming or not!

Reply:
I tried to move the routing logic into the `DefaultRouteDataProvider`. It is possible if I add `RouteTable` to the API in order to trigger route table refresh from the `Router`. 
This doesn't look good because 

- If someone uses custom class that inherits `DefaultRouteDataProvider` in Blazor United application then routing is happening twice. We could make `EndpointRouteDataProvider` public and tell people that they need to inherit from `EndpointRouteDataProvider` in that case. But that is very bad.
- Also `EndpointRouteDataProvider` now has `RouteTable` property that is never used as well as in `GetRouteData` the parameter `string url` is also never used. 

I suggest we leave `RoutingStateProvider` as it was only for the purpose to avoid routing twice for Blazor United. Then in a separate PR create `DefaultRouteDataProvider` that incapsulates the routing logic of `Router` as a solution for the issues that were listed. It will use `RoutingStateProvider` inside to avoid routing twice. @javiercn also mentioned that we could add other things to `RoutingStateProvider` in the future so I think this generic name is good.

Reply:
I tried to move the routing logic into `DefaultRouteDataProvider`. It is possible if I add `RouteTable` to the API to refresh route table from `Router`. This overall looks bad because

- If someone uses custom class that inherits `DefaultRouteDataProvider` in Blazor United application then routing is happening twice. We could make `EndpointRouteDataProvider` public and tell people to inherit from `EndpointRouteDataProvider` in that case but this is bad.
- There is `RouteTable` that is never used in `EndpointRouteDataProvider` as well as in `GetRouteData` parameter `string url` is also not used.

I suggest we use `RoutingStateProvider` only for the purpose to avoid routing twice in `Router`. I will revisit the issues listed previously and try to come up with another solution.